pkg/bgpv1/manager/reconcile.go

```diff
@@ -381,6 +385,210 @@ func (r *ExportPodCIDRReconciler) Reconcile(ctx context.Context, p ReconcilePara
+	// Loop over all existing announcements, delete announcements for pod ip pools that no longer exist.
+	for poolKey := range sc.PodIPPoolAnnouncements {
+		_, found, err := r.poolStore.GetByKey(poolKey)
+		if err != nil {
+			return fmt.Errorf("failed to get pod ip pool from diff store; %w", err)
+		}
+		// If the pod ip pool no longer exists, withdraw all associated routes.
+		if !found {
+			if err := r.withdrawPool(ctx, sc, poolKey); err != nil {
+				return fmt.Errorf("failed to withdraw pod ip pool: %w", err)
+			}
+			continue
+		}
+	}
+
+	// Loop over all pod ip pools, reconcile any updates to the pool.
+	iter := r.poolStore.IterKeys()
+	for iter.Next() {
+		poolKey := iter.Key()
+		pool, found, err := r.poolStore.GetByKey(iter.Key())
+		if err != nil {
+			return fmt.Errorf("failed to get pod ip pool from diff store: %w", err)
+		}
+		if !found {
+			// edgecase: If the pod ip pool was removed between the call to IterKeys() and GetByKey().
+			if err := r.withdrawPool(ctx, sc, poolKey); err != nil {
+				return fmt.Errorf("failed to withdraw pod ip pool: %w", err)
+			}
+			continue
+		}
+
+		r.reconcilePodIPPool(ctx, sc, newc, pool, localPools)
+	}
+	return nil
+}
+
+// withdrawPool removes all announcements for the given pod ip pool.
+func (r *PodIPPoolReconciler) withdrawPool(ctx context.Context, sc *ServerWithConfig, key resource.Key) error {
+	advertisements := sc.PodIPPoolAnnouncements[key]
+	// Loop in reverse order so we can delete without effect to the iteration.
+	for i := len(advertisements) - 1; i >= 0; i-- {
+		advertisement := advertisements[i]
+		if err := sc.Server.WithdrawPath(ctx, types.PathRequest{Path: advertisement}); err != nil {
+			// Persist remaining advertisements
+			sc.PodIPPoolAnnouncements[key] = advertisements
+			return fmt.Errorf("failed to withdraw deleted pod ip pool route: %v: %w", advertisement.NLRI, err)
+		}
+
+		// Delete the advertisement after each withdraw in case we error half way through
+		advertisements = slices.Delete(advertisements, i, i+1)
+	}
+
+	// If all were withdrawn without error, we can delete the whole pod ip pool from the map
+	delete(sc.PodIPPoolAnnouncements, key)
+
+	return nil
+}
+
+// reconcilePodIPPool ensures the CIDRs of the given pool are announced if they are present
+// on the local node, adding missing announcements or withdrawing unwanted ones.
+func (r *PodIPPoolReconciler) reconcilePodIPPool(ctx context.Context,
+	sc *ServerWithConfig,
+	newc *v2alpha1api.CiliumBGPVirtualRouter,
+	pool *v2alpha1api.CiliumPodIPPool,
+	localPools map[string][]netip.Prefix) error {
+	poolKey := resource.NewKey(pool)
+
+	desiredRoutes, err := r.poolDesiredRoutes(newc, pool, localPools)
+	if err != nil {
+		return fmt.Errorf("poolDesiredRoutes(): %w", err)
+	}
+
+	for _, desiredRoute := range desiredRoutes {
+		// If this route has already been announced, don't add it again
+		if slices.IndexFunc(sc.PodIPPoolAnnouncements[poolKey], func(existing *types.Path) bool {
```

Comment:
nit: Wouldn't `slices.ContainsFunc` be a better match here?

Reply:
Commit b18c075 implements this change request.